go/array_string/two_sum/solution.go

```diff
@@ -0,0 +1,23 @@
+// Package solution contains the implementation of
+// the two sum algorithm solution with O(n) complexity
+// and also an amortized O(1) complexity.
+package solution
+
+func twoSum(nums []int, target int) []int {
+	// we need to keep a record of element visited
+	// and their index, to avoid iterating the nums
+	// array multiple times so as to find a match with target
+	visited := make(map[int]int)
+
+	for i, num := range nums {
+		// check if the remaining value is among visited elements
+		if index, ok := visited[target - num]; ok {
+			return []int{index, i}
+		}
+
+		// update the visited record
+		visited[num] = i
+	}
+
+	return []int{}
```

Comment:
It's idiomatic return nil rather than an empty slice. When it's used (e.g. `len()` or `range`) then `nil` functions in the same way, but it doesn't require any allocation.

Reply:
Thanks for the feedback, I'll implement it.